### PR TITLE
added ghapi support

### LINKED ISSUE #384: Use ghapi in GithubDownloader
[ghapi](https://ghapi.fast.ai/) provides an up-to-date, generated interface for the entire GitHub API, complete with pagination support. This library could be utilized in the [`GithubDownloader`](https://github.com/golemfactory/goth/blob/ff44ed782e55cfcb4d44c4030fe19ac398978fcf/goth/runner/download/__init__.py#L37) class which currently relies on manually-built requests to the API.

### PR DESCRIPTION
**Pull request by pradeepbbl**
Resolves #384

### CODE REVIEW

File: goth/runner/download/__init__.py [hunk 1]
```diff
@@ -82,107 +83,35 @@ def _create_cache_dir(self, asset_id: str) -> Path:
         asset_path.mkdir(exist_ok=True, parents=True)
         return asset_path
 
-    def _parse_link_header(self, header_value: str) -> dict:
```
Comment (pradeepbbl) [inline goth/runner/download/__init__.py:85]: replaced this with https://ghapi.fast.ai/page.html#paged, we can also speed up the process by using https://ghapi.fast.ai/page.html#pages (which create multi thread)

File: goth/runner/download/__init__.py [hunk 2]
```diff
@@ -65,7 +66,7 @@ def __init__(
         if purge_cache:
             shutil.rmtree(ASSET_CACHE_DIR)
 
-        self.repo_url = BASE_URL + f"/{owner}/{repo}"
```
Comment (zakaprov) [inline goth/runner/download/__init__.py:68]: `BASE_URL` is no longer used anywhere, you can safely remove it.

File: goth/runner/download/__init__.py [hunk 3]
```diff
@@ -41,8 +41,9 @@ class AssetNotFound(Exception):
 class GithubDownloader(ABC):
     """Base class for downloading assets using GitHub's REST API."""
 
-    repo_url: str
-    """Repo URL to be used as base in API requests."""
+    gh_api: GhApi
+    """GitHub api"""
```
Comment (zakaprov) [inline goth/runner/download/__init__.py:45]:
```suggestion
    """GitHub REST API client."""
```

File: goth/runner/download/__init__.py [hunk 4]
```diff
@@ -82,107 +83,35 @@ def _create_cache_dir(self, asset_id: str) -> Path:
...
-        url = f"{self.repo_url}/actions/workflows"
-        logger.debug("Fetching workflows. url=%s", url)
-        response = self.session.get(url)
-        response.raise_for_status()
-
-        workflows = response.json()["workflows"]
-        logger.debug("workflows=%s", workflows)
-        workflow = next(filter(lambda w: w["name"] == workflow_name, workflows))
+        logger.debug("Fetching workflows. name=%s", workflow_name)
+        workflow = next(filter(lambda w: w["name"] == "CI", self.gh_api.actions.list_repo_workflows().workflows))
```
Comment (zakaprov) [inline goth/runner/download/__init__.py:93]:
```suggestion
        workflows = self.gh_api.actions.list_repo_workflows().workflows
        workflow = next(filter(lambda w: w["name"] == workflow_name, workflows))
```

File: goth/runner/download/__init__.py [hunk 5]
```diff
@@ -286,12 +213,7 @@ def _get_latest_release(
...
         as a substring are considered.
         """
-        url = f"{self.repo_url}/releases"
-        logger.debug("Fetching releases. url=%s", url)
-        response = self.session.get(url)
-        response.raise_for_status()
-
-        all_releases = response.json()
+        all_releases = self.gh_api.repos.list_releases()
         logger.debug("releases=%s", all_releases)
```
Comment (zakaprov) [inline goth/runner/download/__init__.py:217]: This is a minor one, but perhaps it's possible to print this in a different format (e.g. JSON or Python repr)? Currently the debug logs look something like this:
```
2021-03-26 19:24:17,613 DEBUG    goth.runner.download                releases=[- url: https://api.github.com/repos/golemfactory/yagna/releases/40429724
- assets_url: https://api.github.com/repos/golemfactory/yagna/releases/40429724/assets
- upload_url: https://uploads.github.com/repos/golemfactory/yagna/releases/40429724/assets{?name,label}
- html_url: https://github.com/golemfactory/yagna/releases/tag/v0.6.3
- id: 40429724
- author:
  - login: github-actions[bot]
  - id: 41898282
  - node_id: MDM6Qm90NDE4OTgyODI=
  - avatar_url: https://avatars.githubusercontent.com/in/15368?v=4
  - gravatar_id:
  - url: https://api.github.com/users/github-actions%5Bbot%5D
  - html_url: https://github.com/apps/github-actions
  - followers_url: https://api.github.com/users/github-actions%5Bbot%5D/followers
  - following_url: https://api.github.com/users/github-actions%5Bbot%5D/following{/other_user}
  - gists_url: https://api.github.com/users/github-actions%5Bbot%5D/gists{/gist_id}
  - starred_url: https://api.github.com/users/github-actions%5Bbot%5D/starred{/owner}{/repo}
  - subscriptions_url: https://api.github.com/users/github-actions%5Bbot%5D/subscriptions
  - organizations_url: https://api.github.com/users/github-actions%5Bbot%5D/orgs
  - repos_url: https://api.github.com/users/github-actions%5Bbot%5D/repos
  - events_url: https://api.github.com/users/github-actions%5Bbot%5D/events{/privacy}
  - received_events_url: https://api.github.com/users/github-actions%5Bbot%5D/received_events
  - type: Bot
  - site_admin: False
- node_id: MDc6UmVsZWFzZTQwNDI5NzI0
- tag_name: v0.6.3
...
```

You can test this locally by using one of our scripts in the repo, e.g.: `./scripts/download_release.py -v yagna` from the project root.

File: goth/runner/download/__init__.py [hunk 6]
```diff
@@ -82,107 +83,35 @@ def _create_cache_dir(self, asset_id: str) -> Path:
...
+        if commit:
+            paged_workflow_runs = paged(self.gh_api.list_workflow_runs, workflow_id, status="completed")
+
+            for page in paged_workflow_runs:
+                workflow_runs = next(filter(lambda r: r["head_sha"].startswith(commit), page.workflow_runs), None)
+
+                if workflow_runs:
+                    return workflow_runs
+
+        return self.gh_api.list_workflow_runs(workflow_id, branch=branch, status="completed").workflow_runs[0]
```
Comment (zakaprov) [inline goth/runner/download/__init__.py:114]: When using the standalone downloader script from this repo (`./scripts/download_artifacts.py)` from the repo root) this throws an exception for me:
```
Traceback (most recent call last):
  File "./scripts/download_artifacts.py", line 64, in <module>
    downloader.download(
  File "/home/kuba/code/python/goth-ghapi/goth/runner/download/__init__.py", line 193, in download
    latest_run = self._get_latest_run(workflow, branch, commit)
  File "/home/kuba/code/python/goth-ghapi/goth/runner/download/__init__.py", line 126, in _get_latest_run
    return self.gh_api.list_workflow_runs(
  File "/home/kuba/.virtualenvs/ghapi/lib/python3.8/site-packages/ghapi/core.py", line 116, in __getattr__
    def __getattr__(self,k): return self.groups[k] if 'groups' in vars(self) and k in self.groups else stop(AttributeError(k))
  File "/home/kuba/.virtualenvs/ghapi/lib/python3.8/site-packages/fastcore/basics.py", line 202, in stop
    raise e
AttributeError: list_workflow_runs
```
I'll try to find a moment next week to see why this is happening, but I'm mentioning it here if you happen to get there first. :)

**Review by zakaprov — CHANGES_REQUESTED** [review]
Great start, thanks a lot for your contribution! This works properly when run as part of our integration test suite. :)
Most of the changes look fine, but please take a look at my comments.

Also, I just realised our codestyle CI job is not configured to run on pull requests (again, thanks). You'll need to reformat your code before merging. We use `black` for formatting and `flake8` for linting etc.

Let me know if you have any questions.

File: goth/runner/proxy.py [hunk 7]
```diff
@@ -7,7 +7,7 @@
 
 from mitmproxy import options
 import mitmproxy.utils.debug
-from mitmproxy.tools import _main, cmdline, dump
+from mitmproxy.tools import main, cmdline, dump
```
Comment (pradeepbbl) [inline goth/runner/proxy.py:10]: not sure you have this issue but am getting  `ImportError`
```
./scripts/download_release.py -v yagna
Traceback (most recent call last):
  File "/Users/pradeep/Data/Github/goth/./scripts/download_release.py", line 7, in <module>
    from goth.runner.download import (
  File "/Users/pradeep/Data/Github/goth/goth/runner/__init__.py", line 34, in <module>
    from goth.runner.proxy import Proxy, run_proxy
  File "/Users/pradeep/Data/Github/goth/goth/runner/proxy.py", line 10, in <module>
    from mitmproxy.tools import _main, cmdline, dump
ImportError: cannot import name '_main' from 'mitmproxy.tools' (/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/mitmproxy/tools/__init__.py)
```

File: goth/runner/download/__init__.py (hunk 5 above)
Comment (pradeepbbl) [reply goth/runner/download/__init__.py]: converted to JSON string using https://fastcore.fast.ai/xtras.html#obj2dict

File: goth/runner/download/__init__.py (hunk 6 above)
Comment (pradeepbbl) [reply goth/runner/download/__init__.py]: found an issue with [ghapi.actions.download_artifact](https://ghapi.fast.ai/fullapi.html), it's trying to decode the download archive which is breaking. For now, revert back to `request.session`  for download.

ref: https://github.com/fastai/ghapi/issues/22

File: goth/runner/download/__init__.py [hunk 8]
```diff
@@ -82,107 +83,45 @@ def _create_cache_dir(self, asset_id: str) -> Path:
...
-        logger.debug("Fetching workflows. url=%s", url)
-        response = self.session.get(url)
-        response.raise_for_status()
-
-        workflows = response.json()["workflows"]
-        logger.debug("workflows=%s", workflows)
+        logger.debug("Fetching workflows. name=%s", workflow_name)
+        workflows = self.gh_api.actions.list_repo_workflows().workflows
         workflow = next(filter(lambda w: w["name"] == workflow_name, workflows))
         logger.debug("workflow=%s", workflow)
```
Comment (zakaprov) [inline goth/runner/download/__init__.py:95]:
```suggestion
        logger.debug("workflow=%s", json.dumps(obj2dict(workflow)))
```

File: goth/runner/proxy.py (hunk 7 above)
Comment (zakaprov) [reply goth/runner/proxy.py:10]: I just tested this on your branch with `python3.9` on Ubuntu 20.04 and I'm not getting the import error you mentioned. I'd revert these changes as we're primarly targeting Ubuntu 18.04 and 20.04 and both seem to be working fine (the former is running on our GitHub builders).

**Comment by zakaprov** [comment]
btw, let me know if you need my assistance in resolving the conflicts before merging.

**Comment by pradeepbbl** [comment]
nice lint is now running on pull +1

**Comment by pradeepbbl** [comment]
is something wrong with the test suit was running for the last 1hr?

**Comment by zakaprov** [comment]
Yes, I stopped the job and cleaned our test runners. Could you please merge the repo's latest `master` into this branch and try again?

**Comment by pradeepbbl** [comment]
Not sure looks like something wrong with CI rebase to the repo to master and ran both `black` and `flake8` no issues but failing in runners

```
$ git rebase upstream/master
Current branch enhancement/384-use-ghapi is up to date.
$ black  .
All done! ✨ 🍰 ✨
77 files left unchanged.
$ flake8 .
```

**Comment by zakaprov** [comment]
@pradeepbbl I ran `black` locally on your latest branch and I'm getting the same errors as reported by CI. Can you please verify your `black` version? You can do so by running `black --version`. The project is configured to use version `20.8b1`, you should have the same one if you installed it through `poetry` along with `goth`.

Otherwise the PR looks good to go! :)

**Comment by pradeepbbl** [comment]
@zakaprov thanks for the help

**Review by zakaprov — APPROVED** [review]
LGTM! :partying_face: 

Thanks a lot for your contribution! I'll try to tag some more "good first issues" in the repo in case you're interested. :)

**Comment by pradeepbbl** [comment]
you can directly assign to me if you want ;)

**Pull request merged**